internal/unionstore/art/arena.go

```diff
@@ -0,0 +1,370 @@
+		data []byte
+	)
+	if len(f.nodeAllocator.freeNode16) > 0 {
+		addr = f.nodeAllocator.freeNode16[len(f.nodeAllocator.freeNode16)-1]
+		f.nodeAllocator.freeNode16 = f.nodeAllocator.freeNode16[:len(f.nodeAllocator.freeNode16)-1]
+		data = f.nodeAllocator.getData(addr)
+	} else {
+		addr, data = f.nodeAllocator.alloc(node16size, true)
+	}
+	n16 := (*node16)(unsafe.Pointer(&data[0]))
+	n16.init()
+	return addr, n16
+}
+
+func (f *artAllocator) freeNode16(addr nodeAddr) {
+	f.nodeAllocator.freeNode16 = append(f.nodeAllocator.freeNode16, addr)
+}
+
+func (f *artAllocator) getNode16(addr nodeAddr) *node16 {
+	if addr.isNull() {
+		return nil
+	}
+	data := f.nodeAllocator.getData(addr)
+	return (*node16)(unsafe.Pointer(&data[0]))
+}
+
+func (f *artAllocator) allocNode48() (nodeAddr, *node48) {
+	var (
+		addr nodeAddr
+		data []byte
+	)
+	if len(f.nodeAllocator.freeNode48) > 0 {
+		addr = f.nodeAllocator.freeNode48[len(f.nodeAllocator.freeNode48)-1]
+		f.nodeAllocator.freeNode48 = f.nodeAllocator.freeNode48[:len(f.nodeAllocator.freeNode48)-1]
+		data = f.nodeAllocator.getData(addr)
+	} else {
+		addr, data = f.nodeAllocator.alloc(node48size, true)
+	}
+	n48 := (*node48)(unsafe.Pointer(&data[0]))
+	n48.init()
+	return addr, n48
+}
+
+func (f *artAllocator) freeNode48(addr nodeAddr) {
+	f.nodeAllocator.freeNode48 = append(f.nodeAllocator.freeNode48, addr)
+}
+
+func (f *artAllocator) getNode48(addr nodeAddr) *node48 {
+	if addr.isNull() {
+		return nil
+	}
+	data := f.nodeAllocator.getData(addr)
+	return (*node48)(unsafe.Pointer(&data[0]))
+}
+
+func (f *artAllocator) allocNode256() (nodeAddr, *node256) {
+	var (
+		addr nodeAddr
+		data []byte
+	)
+	if len(f.nodeAllocator.freeNode256) > 0 {
+		addr = f.nodeAllocator.freeNode256[len(f.nodeAllocator.freeNode256)-1]
+		f.nodeAllocator.freeNode256 = f.nodeAllocator.freeNode256[:len(f.nodeAllocator.freeNode256)-1]
+		data = f.nodeAllocator.getData(addr)
+	} else {
+		addr, data = f.nodeAllocator.alloc(node256size, true)
+	}
+	n256 := (*node256)(unsafe.Pointer(&data[0]))
+	n256.init()
+	return addr, n256
+}
+
+func (f *artAllocator) freeNode256(addr nodeAddr) {
+	f.nodeAllocator.freeNode256 = append(f.nodeAllocator.freeNode256, addr)
+}
+
+func (f *artAllocator) getNode256(addr nodeAddr) *node256 {
+	if addr.isNull() {
+		return nil
+	}
+	data := f.nodeAllocator.getData(addr)
+	return (*node256)(unsafe.Pointer(&data[0]))
+}
+
+func (f *artAllocator) allocLeaf(key Key) (nodeAddr, *leaf) {
+	size := leafSize + len(key)
+	addr, data := f.nodeAllocator.alloc(size, true)
+	lf := (*leaf)(unsafe.Pointer(&data[0]))
+	lf.klen = uint16(len(key))
+	lf.flags = 0
+	lf.vAddr = nullAddr
+	copy(data[leafSize:], key)
+	return addr, lf
+}
+
+func (f *artAllocator) getLeaf(addr nodeAddr) *leaf {
+	if addr.isNull() {
+		return nil
+	}
+	data := f.nodeAllocator.getData(addr)
+	return (*leaf)(unsafe.Pointer(&data[0]))
+}
+
+// memArena get all the data, DO NOT access others data.
+func (a *memArena) getData(addr nodeAddr) []byte {
```

Comment:
The usage of this function is always converted to a pointer. So I suppose the function does not have to exist? We may have a direct function converting address to pointers.